<a href="https://colab.research.google.com/github/arvynathaniel/Python/blob/main/Mental_Health_Data_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**Mental Health Survey Data Cleaning Project**

In this project, we will be looking at the 'Mental Health Survey' dataset. The main goal of this project is to take a look at the data presented by this dataset and then clean it, so that the cleaned data can further be used on the next exploratory data analysis (EDA) project.

The main work sequence that will be performed in this project:
1.   Calling in the libraries and dataset
2.   Performing a light data exploration on the dataset
3.   Cleaning the data and performing some feature engineering if needed
4.   Exporting the dataset into a new csv file

Our thanks to the provider of this dataset:

https://www.kaggle.com/datasets/osmi/mental-health-in-tech-survey


##**1. Calling in the libraries and dataset**

###1a. Libraries

In [466]:
# pandas to help us performing actions on the dataset in a tabular form
import pandas as pd
import numpy as np

###1b. Dataset

In [467]:
data = pd.read_csv('Mental Health Survey.csv')

##**2. Light Data Exploration**

###2a. General overview

In [468]:
data.head()

Timestamp  Age  Gender         Country state self_employed  \
0  2014-08-27 11:29:31  37   Female  United States   IL    NaN            
1  2014-08-27 11:29:37  44   M       United States   IN    NaN            
2  2014-08-27 11:29:44  32   Male    Canada          NaN   NaN            
3  2014-08-27 11:29:46  31   Male    United Kingdom  NaN   NaN            
4  2014-08-27 11:30:22  31   Male    United States   TX    NaN            

  family_history treatment work_interfere    no_employees  ...  \
0  No             Yes       Often          6-25            ...   
1  No             No        Rarely         More than 1000  ...   
2  No             No        Rarely         6-25            ...   
3  Yes            Yes       Often          26-100          ...   
4  No             No        Never          100-500         ...   

                leave mental_health_consequence phys_health_consequence  \
0  Somewhat easy       No                        No                       
1  Don't know          Maybe                     No                       
2  Somewhat difficult  No                        No                       
3  Somewhat difficult  Yes                       Yes                      
4  Don't know          No                        No                       

      coworkers supervisor mental_health_interview phys_health_interview  \
0  Some of them  Yes        No                      Maybe                  
1  No            No         No                      No                     
2  Yes           Yes        Yes                     Yes                    
3  Some of them  No         Maybe                   Maybe                  
4  Some of them  Yes        Yes                     Yes                    

  mental_vs_physical obs_consequence comments  
0  Yes                No              NaN      
1  Don't know         No              NaN      
2  No                 No              NaN      
3  No                 Yes             NaN      
4  Don't know         No              NaN      

[5 rows x 27 columns]

In [469]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1259 entries, 0 to 1258
Data columns (total 27 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Timestamp                  1259 non-null   object
 1   Age                        1259 non-null   int64 
 2   Gender                     1259 non-null   object
 3   Country                    1259 non-null   object
 4   state                      744 non-null    object
 5   self_employed              1241 non-null   object
 6   family_history             1259 non-null   object
 7   treatment                  1259 non-null   object
 8   work_interfere             995 non-null    object
 9   no_employees               1259 non-null   object
 10  remote_work                1259 non-null   object
 11  tech_company               1259 non-null   object
 12  benefits                   1259 non-null   object
 13  care_options               1259 non-null   object
 14  wellness

Regarding the feature names of the dataset, the provider of this dataset previously gave us a clue about what data each feature represent. We will be displaying them in a Pandas dataframe format so that it is easier for us to read.

In [470]:
# Storing all the feature description in a variable
RawFeatures = 'Timestamp;Age;Gender;Country;state: If you live in the United States,which state or territory do you live in?;self_employed: Are you self-employed?;family_history: Do you have a family history of mental illness?;treatment: Have you sought treatment for a mental health condition?;work_interfere: If you have a mental health condition,do you feel that it interferes with your work?;no_employees: How many employees does your company or organization have?;remote_work: Do you work remotely (outside of an office) at least 50% of the time?;tech_company: Is your employer primarily a tech company/organization?;benefits: Does your employer provide mental health benefits?;care_options: Do you know the options for mental health care your employer provides?;wellness_program: Has your employer ever discussed mental health as part of an employee wellness program?;seek_help: Does your employer provide resources to learn more about mental health issues and how to seek help?;anonymity: Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources?;leave: How easy is it for you to take medical leave for a mental health condition?;mentalhealthconsequence: Do you think that discussing a mental health issue with your employer would have negative consequences?;physhealthconsequence: Do you think that discussing a physical health issue with your employer would have negative consequences?;coworkers: Would you be willing to discuss a mental health issue with your coworkers?;supervisor: Would you be willing to discuss a mental health issue with your direct supervisor(s)?;mentalhealthinterview: Would you bring up a mental health issue with a potential employer in an interview?;physhealthinterview: Would you bring up a physical health issue with a potential employer in an interview?;mentalvsphysical: Do you feel that your employer takes mental health as seriously as physical health?;obs_consequence: Have you heard of or observed negative consequences for coworkers with mental health conditions in your workplace?;comments: Any additional notes or comments'

# Splitting each feature into a separate item in a list
Features = RawFeatures.split(';')

# Separating each feature and its description into two separate lists
ListFeature = []
ListDesc = []
for i in Features:
    part1 = i.partition(':')[0]
    part2 = i.partition(':')[2]
    ListFeature.append(part1)
    ListDesc.append(part2)

# Creating the description dataframe
FeaturesTable = pd.DataFrame(list(zip(ListFeature, ListDesc))) # Creating the dataframe
pd.set_option("display.max_colwidth", 0)                       # Displaying the description's column in a full length
FeaturesTable.columns = ['Feature', 'Description']             # Giving each dataframe's column a name

# Displaying the description dataframe
print('Feature Description')
FeaturesTable

Feature Description


Feature  \
0   Timestamp                 
1   Age                       
2   Gender                    
3   Country                   
4   state                     
5   self_employed             
6   family_history            
7   treatment                 
8   work_interfere            
9   no_employees              
10  remote_work               
11  tech_company              
12  benefits                  
13  care_options              
14  wellness_program          
15  seek_help                 
16  anonymity                 
17  leave                     
18  mentalhealthconsequence   
19  physhealthconsequence     
20  coworkers                 
21  supervisor                
22  mentalhealthinterview     
23  physhealthinterview       
24  mentalvsphysical          
25  obs_consequence           
26  comments                  

                                                                                                              Description  
0                                                                                                                          
1                                                                                                                          
2                                                                                                                          
3                                                                                                                          
4    If you live in the United States,which state or territory do you live in?                                             
5    Are you self-employed?                                                                                                
6    Do you have a family history of mental illness?                                                                       
7    Have you sought treatment for a mental health condition?                                                              
8    If you have a mental health condition,do you feel that it interferes with your work?                                  
9    How many employees does your company or organization have?                                                            
10   Do you work remotely (outside of an office) at least 50% of the time?                                                 
11   Is your employer primarily a tech company/organization?                                                               
12   Does your employer provide mental health benefits?                                                                    
13   Do you know the options for mental health care your employer provides?                                                
14   Has your employer ever discussed mental health as part of an employee wellness program?                               
15   Does your employer provide resources to learn more about mental health issues and how to seek help?                   
16   Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources?  
17   How easy is it for you to take medical leave for a mental health condition?                                           
18   Do you think that discussing a mental health issue with your employer would have negative consequences?               
19   Do you think that discussing a physical health issue with your employer would have negative consequences?             
20   Would you be willing to discuss a mental health issue with your coworkers?                                            
21   Would you be willing to discuss a mental health issue with your direct supervisor(s)?                                 
22   Would you bring up a mental health issue with a potential employer in an interview?                                   
23   Would you bring up a physical health issue with a potential employer in an interview?                                 
24   Do you feel that your employer takes mental he

###2b. Missing values

In [471]:
# Calculating the missing values for each feature
MissingValues = data.isnull().sum()

# Turning the previous output into a dataframe format
MissingValues = pd.DataFrame(MissingValues)
MissingValues = MissingValues.reset_index()
MissingValues.columns = ['Feature', 'Count']

# Calculating each feature's missing values percentage
ListMissingVPerc = []
for i in MissingValues['Count']:
    missingperc = i/1259 * 100 # Total entries = 1259 (from previous step)
    ListMissingVPerc.append(missingperc)
ListMissingVPerc = [round(num, 2) for num in ListMissingVPerc]

# Adding the missing values percentage column to the report
MissingValues['Percentage (%)'] = ListMissingVPerc

# Printing the missing values report
print('Missing values on each feature: ')
print(MissingValues.to_string(index = False))

Missing values on each feature: 
                  Feature  Count  Percentage (%)
                Timestamp      0            0.00
                      Age      0            0.00
                   Gender      0            0.00
                  Country      0            0.00
                    state    515           40.91
            self_employed     18            1.43
           family_history      0            0.00
                treatment      0            0.00
           work_interfere    264           20.97
             no_employees      0            0.00
              remote_work      0            0.00
             tech_company      0            0.00
                 benefits      0            0.00
             care_options      0            0.00
         wellness_program      0            0.00
                seek_help      0            0.00
                anonymity      0            0.00
                    leave      0            0.00
mental_health_consequence      0    

###2c. Duplicated values

In [472]:
DuplicatedValues = data.duplicated().sum()
print('Duplicated data : ' + str(DuplicatedValues))

Duplicated data : 0


###2d. Unique values

In [473]:
# Getting the feature names in a list
DataCols = list(data.columns.values)

# Calculating the number of each feature's unique values
ListUniqueC = []
for i in data.columns:
    UCount = len(data[i].unique())
    ListUniqueC.append(UCount)

# Creating the unique values table
UniqueCTable = pd.DataFrame(list(zip(DataCols, ListUniqueC)))
UniqueCTable.columns = ['Feature', 'Unique Values']

# Printing the report
print('Unique values in each feature:')
print(UniqueCTable.to_string(index = False))

Unique values in each feature:
                  Feature  Unique Values
                Timestamp           1246
                      Age             53
                   Gender             49
                  Country             48
                    state             46
            self_employed              3
           family_history              2
                treatment              2
           work_interfere              5
             no_employees              6
              remote_work              2
             tech_company              2
                 benefits              3
             care_options              3
         wellness_program              3
                seek_help              3
                anonymity              3
                    leave              5
mental_health_consequence              3
  phys_health_consequence              3
                coworkers              3
               supervisor              3
  mental_health_interview 

##**3. Data Cleaning and Feature Engineering**

###Timestamp

The 'Timestamp' feature does not really have any meaning or importance in this dataset - it does not indicate a time when a person is doing the survey. We will drop this feature.

In [474]:
# Dropping the 'Timestamp' feature
data.drop('Timestamp', axis = 1, inplace = True)

# Checking the features of the dataset
print(data.columns)

Index(['Age', 'Gender', 'Country', 'state', 'self_employed', 'family_history',
       'treatment', 'work_interfere', 'no_employees', 'remote_work',
       'tech_company', 'benefits', 'care_options', 'wellness_program',
       'seek_help', 'anonymity', 'leave', 'mental_health_consequence',
       'phys_health_consequence', 'coworkers', 'supervisor',
       'mental_health_interview', 'phys_health_interview',
       'mental_vs_physical', 'obs_consequence', 'comments'],
      dtype='object')


###Age

We will check if there are any entries with abnormal 'Age' value. By knowing that the target of the dataset are people who work in tech industry, it can be assumed that the youngest of the population should be around 18 (US legal age) and the oldest around 80 (still possible).

In [475]:
# Storing the age feature in a variable
Age = data['Age']

# Age feature range
AgeMin = min(Age)
AgeMax = max(Age)

# Entries with abnormal age value count
AgeLO = (Age < 18).sum()
AgeUO = (Age > 80).sum()
AgeOCount = AgeLO + AgeUO
AgeOCountP = AgeOCount / 1259

# Displaying the information
print('Highest age in the entries : ' + str(AgeMax))
print('Lowest age in the entries  : ' + str(AgeMin))
print('People younger than 18     : ' + str(AgeLO) + ' people')
print('People older than 80       : ' + str(AgeUO) + ' people')
print('Abnormal entries count     : ' + str(AgeOCount))
print('Abnormal entries percentage: ' + str(AgeOCountP))

Highest age in the entries : 99999999999
Lowest age in the entries  : -1726
People younger than 18     : 6 people
People older than 80       : 2 people
Abnormal entries count     : 8
Abnormal entries percentage: 0.006354249404289118


Luckily, there are only 8 entries with an abnormal 'Age' value, accounting for around 0.64% of the total entries. We need to replace these values with a value that is more likely to occur. A simple method to do this is to re-input them with the mean of the 'Age' feature. However, since the mean value is affected by the highly abnormal outlier (99999999999), we will instead use the median value for this process. Let us first check the what the median value of the 'Age' feature is.



In [476]:
AgeMedian = data['Age'].median()
print('Median of age feature : ' + str(AgeMedian))

Median of age feature : 31.0


The median looks quite normal and can be used to impute for the abnormal values.

In [477]:
# Replacing the age value of the abnormal age entries
data['Age'].mask((data['Age'] < 18) | (data['Age'] > 90),
                 31,
                 inplace = True)

# Checking if the values are replaced correctly
Age2 = data['Age']
AgeMin2 = min(Age2)
AgeMax2 = max(Age2)
AgeCount2 = len(Age2)
print('Lowest age in the entries       : ' + str(AgeMin2))
print('Highest age in the entries      : ' + str(AgeMax2))
print('Total counts of the age entries : ' + str(AgeCount2))

Lowest age in the entries       : 18
Highest age in the entries      : 72
Total counts of the age entries : 1259


###Gender

There are 49 unique values in the 'Gender' feature. We will first check what these unique values are before deciding what to do with this feature.

In [478]:
Gender = data['Gender']
GenderUV = Gender.unique()
print(GenderUV)

['Female' 'M' 'Male' 'male' 'female' 'm' 'Male-ish' 'maile' 'Trans-female'
 'Cis Female' 'F' 'something kinda male?' 'Cis Male' 'Woman' 'f' 'Mal'
 'Male (CIS)' 'queer/she/they' 'non-binary' 'Femake' 'woman' 'Make' 'Nah'
 'All' 'Enby' 'fluid' 'Genderqueer' 'Female ' 'Androgyne' 'Agender'
 'cis-female/femme' 'Guy (-ish) ^_^' 'male leaning androgynous' 'Male '
 'Man' 'Trans woman' 'msle' 'Neuter' 'Female (trans)' 'queer'
 'Female (cis)' 'Mail' 'cis male' 'A little about you' 'Malr' 'p' 'femail'
 'Cis Man' 'ostensibly male, unsure what that really means']


From the list above, it seems that there was no some sort of data validation process for when the subjects made the 'Gender' feature entry. For the sake of simplicity, we will do some feature engineering and categorize the 'Gender' feature into only three values: 'Male', 'Female', and 'Other'.

In [479]:
# Defining the Male and Female categories
male = ['M', 'Male', 'male', 'm', 'maile', 'Mal', 'Make', 'Male ', 'Man', 'msle', 'Mail', 'Malr']
female = ['Female', 'female', 'F', 'Woman', 'f', 'Femake', 'woman', 'Female ', 'femail']

# Defining the Other category
other1 = [x for x in GenderUV if x not in male]
other2 = [x for x in other1 if x not in female]

# Replacing the values in Gender feature into three categories only
data['Gender'].replace(to_replace = male, value = 'Male', inplace=True)
data['Gender'].replace(to_replace = female, value = 'Female', inplace=True)
data['Gender'].replace(to_replace = other2, value = 'Other', inplace=True)

# Displaying the report
print('Number of unique values in Gender feature : ' + 
      str(len(data['Gender'].unique())))
print('Unique values in Gender feature           : ' + 
      str(data['Gender'].unique()))

Number of unique values in Gender feature : 3
Unique values in Gender feature           : ['Female' 'Male' 'Other']


###Country

In [480]:
# Storing each unique value of Country in a list
Country = data['Country']
CountryUV = Country.unique()

# Storing the count of each unique value of Country in a list
ListCountryC = []
for i in CountryUV:
    CountryC = (data['Country'] == i).sum()
    ListCountryC.append(CountryC)

# Displaying the report in a table
TableCountry = pd.DataFrame(list(zip(CountryUV, ListCountryC)))
TableCountry.columns = ['Country', 'Count']
print("Unique values in 'Country' and each of its count:")
print(TableCountry.
      sort_values(['Count', 'Country'], 
                  ascending = [0,1]).
      to_string(index = False))

Unique values in 'Country' and each of its count:
               Country  Count
         United States    751
        United Kingdom    185
                Canada     72
               Germany     45
               Ireland     27
           Netherlands     27
             Australia     21
                France     13
                 India     10
           New Zealand      8
                 Italy      7
                Poland      7
                Sweden      7
           Switzerland      7
               Belgium      6
                Brazil      6
          South Africa      6
                Israel      5
              Bulgaria      4
             Singapore      4
               Austria      3
               Finland      3
                Mexico      3
                Russia      3
              Colombia      2
               Croatia      2
               Denmark      2
                Greece      2
              Portugal      2
          Bahamas, The      1
Bosnia and Herzegovi

We will be doing a lot of unique values counting in this data cleaning process for the next features, so it will be more efficient if we define a function to do this instead of repeating the tedious longer codes. The following function is to get the unique values and each of their counts in a discrete feature.

In [481]:
def uniquevaluescount(data, feature):
    # Storing each unique value of feature in a list
    features = data[feature]
    listUV = features.unique()
    listUV = [i for i in listUV if type(i) is not float]

    # Storing the count of each unique value of feature in a list
    listUVcount = []
    for i in listUV:
        UVcount = (data[feature] == i).sum()
        listUVcount.append(UVcount)
    
    # Displaying the report in a table
    table = pd.DataFrame(list(zip(listUV, listUVcount)))
    table.columns = [feature, 'Count']
    print(table.
          sort_values(['Count', feature], 
                      ascending = [0,1]).
          to_string(index = False))
    print('Total : ' + str(sum(listUVcount)))


###State

In the 'state' feature, the number of missing value is quite large, accounting for 40.91% of the total number of entries. Before deciding what to do with the missing values, we first need to check a possible explanation as to why the missing value is that large.

'state' is applicable for only certain countries that have states in them, for example: United States that has 50 states. For countries that have no states, then it is most likely that their 'state' feature is empty. Let us check if this is the case.

In [482]:
# Count: Entries outside US
NotUS = (data['Country'] != 'United States').sum()

# Count: Entries outside US that have empty state feature
NotUSNullState = data[data['Country'] != 'United States']['state'].isnull().sum()

# Count: Entries in US
US = (data['Country'] == 'United States').sum()

# Count: Entries in US that have empty state feature
USNullState = data[data['Country'] == 'United States']['state'].isnull().sum()

# Total entries with empty state
NullState = data['state'].isnull().sum()

# Displaying the report
print('Entries Outside US                  : ' + str(NotUS))
print('Entries Outside US with empty state : ' + str(NotUSNullState))
print('Entries In US                       : ' + str(US))
print('Entries In US with empty state      : ' + str(USNullState))
print('Total entries with empty state      : ' + str(NullState))

Entries Outside US                  : 508
Entries Outside US with empty state : 504
Entries In US                       : 751
Entries In US with empty state      : 11
Total entries with empty state      : 515


From the report above, it can be seen that most of the non-empty 'state' feature only occur in the entries with 'Country' value 'United States'. There are only 4 entries that are outside of 'United States' that have non-empty 'state' feature. When we are specifically analyzing the data inside 'United States', this 'state' feature can be pretty useful. Because of this, we will not discard the entire 'state' feature despite the large missing values, but will instead do a little cleaning on this feature. 

####Inside of United States

Let us check the unique values in the 'state' feature of the US 'Country' entries.

In [483]:
uniquevaluescount(data[data['Country'] == 'United States'], 'state')

state  Count
   CA    138
   WA     70
   NY     56
   TN     45
   TX     44
   OH     30
   OR     29
   PA     29
   IL     28
   IN     27
   MI     22
   MN     21
   MA     20
   FL     15
   NC     14
   VA     14
   GA     12
   MO     12
   WI     12
   UT     10
   CO      9
   AL      8
   AZ      7
   MD      7
   NJ      6
   OK      6
   KY      5
   SC      5
   CT      4
   DC      4
   IA      4
   KS      3
   NH      3
   NV      3
   SD      3
   VT      3
   NE      2
   NM      2
   WY      2
   ID      1
   LA      1
   ME      1
   MS      1
   RI      1
   WV      1
Total : 740


There seems no problem with the unique values: no typo or any abnormal format. Now, let us count the count percentage of the missing values.

In [484]:
# Calculating the 'state' missing value percentage of the 'United States' entries
USNullStatePerc = USNullState / US * 100
USNullStatePerc = "{:.2f}".format(USNullStatePerc)
print("Entries in US with Null 'state' : " + str(USNullStatePerc) + '%')

Entries in US with Null 'state' : 1.46%


In this opportunity, we will also define a function to give a report about the missing value percentage in a feature, which will be helpful in the following data cleaning process.

In [485]:
def missingvaluepercentage(data, feature):
    # Count of missing value
    missingvaluecount = data[feature].isnull().sum()
    
    # Count of total data
    totalcount = len(data[feature])

    # Missing value percentage calculation
    missingvalue = missingvaluecount / totalcount * 100
    missingvalue = "{:.2f}".format(missingvalue)

    # Displaying the misisng value report
    print('Missing value : ' + str(missingvalue) + '%')

The percentage of missing values in the 'state' feature of the entries in 'United States' is relatively small, around 1.46% and still can be used for further analysis. We will just rename the missing values with 'No Data' and see if further deeper analysis can extract some information out of it.

In [486]:
# Replacing the 'state' missing values of the 'United States' entries with 'No Data'
data.loc[data.Country == 'United States', 'state'] = data.loc[data.Country == 'United States', 'state'].fillna('No Data')

# Checking if the missing values have been replaced correctly
USNoDataState = len(data[data['Country'] == 'United States'][data['state'] == 'No Data'])
USNullState = data[data['Country'] == 'United States']['state'].isnull().sum()
NullState = data['state'].isnull().sum()

# Displaying the report
print("Entries In US With 'No Data' State : " + str(USNoDataState))
print('Entries in US with empty state     : ' + str(USNullState))
print('Total entries with empty state     : ' + str(NullState))
print('Entries Outside US                 : ' + str(NotUS))

Entries In US With 'No Data' State : 11
Entries in US with empty state     : 0
Total entries with empty state     : 504
Entries Outside US                 : 508


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """


The 'state' missing values of the 'United States' entries have been successfully replaced with 'No Data'. We will no proceed to the entries outside of the 'United States'.

####Outside of United States

Let us check the unique values in the 'state' feature of entries outside of the US 'Country'.

In [487]:
uniquevaluescount(data[data['Country'] != 'United States'], 'state')

state  Count
   IL      1
   MD      1
   NY      1
   UT      1
Total : 4


In [488]:
# Getting the unique 'state' values of entries outside 'United States'
liststateOutsideUS = list(data[data['Country'] != 'United States']['state'].unique())

# Omitting the nan numerical value
liststateOutsideUS = [i for i in liststateOutsideUS if type(i) is not float]

# Getting the 'Country' for each of the 'state'
listcountryOutsideUS = []
for i in liststateOutsideUS:
    countryOutsideUS = data[data['Country'] != 'United States'].loc[data['state'] == i, 'Country'].iloc[0]
    listcountryOutsideUS.append(str(countryOutsideUS))

# Displaying the report
Table = pd.DataFrame(list(zip(liststateOutsideUS, listcountryOutsideUS)))
Table.columns = ['State', 'Country']
print("Unique 'state' value of entries outside of 'United States' and their corresponding 'Country :'")
print(Table.to_string(index = False))

Unique 'state' value of entries outside of 'United States' and their corresponding 'Country :'
State      Country
   NY       Latvia
   MD       Israel
   IL Bahamas, The
   UT     Bulgaria


Each 'state' resides on different countries. In further analysis, these 'state' values might not be used. We will just leave them like that for now.

###self_employed

Checking the unique values in the 'self_employed' feature.

In [489]:
# Checking the unique values in the 'self_employed' feature
uniquevaluescount(data, 'self_employed')

self_employed  Count
           No   1095
          Yes    146
Total : 1241


There seems no problem with the unique values: no typo or any abnormal format. Now, let us count the count percentage of the missing values.

In [490]:
missingvaluepercentage(data, 'self_employed')

Missing value : 1.43%


The percentage of missing values in the 'self_employed' feature is relatively small, around 1.43% and still can be used for further analysis. We will try filling in the missing values randomly based on the probability of each 'yes' or 'no' value.

In [491]:
# Defining a function to calculate the probability of each unique value in a feature
def valueprobability(data, feature):
    # Storing each unique value of feature in a list
    features = data[feature]
    global listUV
    listUV = features.unique()
    listUV = [i for i in listUV if type(i) is not float]

    # Storing the count of each unique value of feature in a list
    global listUVcount
    listUVcount = []
    for i in listUV:
        UVcount = (data[feature] == i).sum()
        listUVcount.append(UVcount)
    
    # Calculating the total count of the feature
    totalcount = sum(listUVcount)

    # Calculating the probability of each unique value in the feature
    global listUVprob
    listUVprob = []
    for i in listUVcount:
        UVprob = i / totalcount
        listUVprob.append(UVprob)
    
    # Displaying the report in a table
    table = pd.DataFrame(list(zip(listUV, listUVcount, listUVprob)))
    table.columns = [feature, 'Count', 'Probabiity']
    print(table.
          sort_values(['Count', feature], 
                      ascending = [0,1]).
          to_string(index = False))
    

In [492]:
# Calculating the probability of each value
valueprobability(data, 'self_employed')

self_employed  Count  Probabiity
           No   1095    0.882353
          Yes    146    0.117647


Filling in the missing values of the 'self_employed' feature based on the probability of each unique value in the feature.

In [493]:
# Filling in the missing values
data['self_employed'] = data['self_employed'].fillna(pd.Series(np.random.choice(listUV,
                                                                               p = listUVprob,
                                                                               size = len(data))))

Checking if the missing values have been filled in, then the count of unique values of the updated feature.

In [494]:
# Checking if the missing values have been filled in
print(missingvaluepercentage(data, 'self_employed'))

Missing value : 0.00%
None


In [495]:
# Checking the unique values in the 'self_employed' feature
print(uniquevaluescount(data, 'self_employed'))

self_employed  Count
           No   1109
          Yes    150
Total : 1259
None


###family_history

Checking the unique values in the 'family_history' feature.

In [496]:
# Checking the unique values in the 'self_employed' feature
uniquevaluescount(data, 'family_history')

family_history  Count
            No    767
           Yes    492
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###treatment

Checking the unique values in the 'treatment' feature.

In [497]:
# Checking the unique values in the 'self_employed' feature
uniquevaluescount(data, 'treatment')

treatment  Count
      Yes    637
       No    622
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###work_interfere

Checking the unique values in the 'work_interfere' feature.

In [498]:
# Checking the unique values in the 'self_employed' feature
uniquevaluescount(data, 'work_interfere')

work_interfere  Count
     Sometimes    465
         Never    213
        Rarely    173
         Often    144
Total : 995


There seems no problem with the unique values: no typo or any abnormal format. However, there are some entries with missing values that we must deal with.

In [499]:
missingvaluepercentage(data, 'work_interfere')

Missing value : 20.97%


The missing values in the 'work_interfere' feature is quite large. However, since this feature may present a rather important insight when we analyze the dataset further, we will try to fill in the missing values instead of dropping the entire feature. 

In [500]:
# Calculating the probability of each value
valueprobability(data, 'work_interfere')

work_interfere  Count  Probabiity
     Sometimes    465    0.467337
         Never    213    0.214070
        Rarely    173    0.173869
         Often    144    0.144724


In [501]:
# Filling in the missing values based on the probability of each unique value
data['work_interfere'] = data['work_interfere'].fillna(pd.Series(np.random.choice(listUV,
                                                                                  p = listUVprob,
                                                                                  size = len(data))))

Checking if the missing values have been filled in, then the count of unique values of the updated feature.

In [502]:
# Checking if the missing values have been filled in
print(missingvaluepercentage(data, 'work_interfere'))

Missing value : 0.00%
None


In [503]:
# Checking the unique values in the 'self_employed' feature
print(uniquevaluescount(data, 'work_interfere'))

work_interfere  Count
     Sometimes    597
         Never    271
        Rarely    214
         Often    177
Total : 1259
None


###no_employees

Checking the unique values in the 'no_employees' feature.

In [504]:
# Checking the unique values in the 'no_employees' feature
uniquevaluescount(data, 'no_employees')

  no_employees  Count
          6-25    290
        26-100    289
More than 1000    282
       100-500    176
           1-5    162
      500-1000     60
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format. However, we will just rename the unique value 'More than 1000' to '>1000' instead to make it shorter.

In [505]:
data['no_employees'] = data['no_employees'].replace(['More than 1000'], '>1000')

In [506]:
# Checking the unique values in the 'no_employees' feature
uniquevaluescount(data, 'no_employees')

no_employees  Count
        6-25    290
      26-100    289
       >1000    282
     100-500    176
         1-5    162
    500-1000     60
Total : 1259


###remote_work

Checking the unique values in the 'remote_work' feature.

In [507]:
# Checking the unique values in the 'remote_work' feature
uniquevaluescount(data, 'remote_work')

remote_work  Count
         No    883
        Yes    376
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###tech_company

Checking the unique values in the 'tech_company' feature.

In [508]:
# Checking the unique values in the 'tech_company' feature
uniquevaluescount(data, 'tech_company')

tech_company  Count
         Yes   1031
          No    228
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###benefits

Checking the unique values in the 'benefits' feature.

In [509]:
# Checking the unique values in the 'benefits' feature
uniquevaluescount(data, 'benefits')

  benefits  Count
       Yes    477
Don't know    408
        No    374
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###care_options

Checking the unique values in the 'care_options' feature.

In [510]:
# Checking the unique values in the 'care_options' feature
uniquevaluescount(data, 'care_options')

care_options  Count
          No    501
         Yes    444
    Not sure    314
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###wellness_program

Checking the unique values in the 'wellness_program' feature.

In [511]:
# Checking the unique values in the 'wellness_program' feature
uniquevaluescount(data, 'wellness_program')

wellness_program  Count
              No    842
             Yes    229
      Don't know    188
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###seek_help

Checking the unique values in the 'seek_help' feature.

In [512]:
# Checking the unique values in the 'seek_help' feature
uniquevaluescount(data, 'seek_help')

 seek_help  Count
        No    646
Don't know    363
       Yes    250
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###anonymity

Checking the unique values in the 'anonymity' feature.

In [513]:
# Checking the unique values in the 'anonymity' feature
uniquevaluescount(data, 'anonymity')

 anonymity  Count
Don't know    819
       Yes    375
        No     65
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###leave

Checking the unique values in the 'leave' feature.

In [514]:
# Checking the unique values in the 'leave' feature
uniquevaluescount(data, 'leave')

             leave  Count
        Don't know    563
     Somewhat easy    266
         Very easy    206
Somewhat difficult    126
    Very difficult     98
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###mental_health_consequence

Checking the unique values in the 'mental_health_consequence' feature.

In [515]:
# Checking the unique values in the 'mental_health_consequence' feature
uniquevaluescount(data, 'mental_health_consequence')

mental_health_consequence  Count
                       No    490
                    Maybe    477
                      Yes    292
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###phys_health_consequence

Checking the unique values in the 'phys_health_consequence' feature.

In [516]:
# Checking the unique values in the 'phys_health_consequence' feature
uniquevaluescount(data, 'phys_health_consequence')

phys_health_consequence  Count
                     No    925
                  Maybe    273
                    Yes     61
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###coworkers

Checking the unique values in the 'coworkers' feature.

In [517]:
# Checking the unique values in the 'coworkers' feature
uniquevaluescount(data, 'coworkers')

   coworkers  Count
Some of them    774
          No    260
         Yes    225
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###supervisor

Checking the unique values in the 'supervisor' feature.

In [518]:
# Checking the unique values in the 'supervisor' feature
uniquevaluescount(data, 'supervisor')

  supervisor  Count
         Yes    516
          No    393
Some of them    350
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###mental_health_interview

Checking the unique values in the 'mental_health_interview' feature.

In [519]:
# Checking the unique values in the 'mental_health_interview' feature
uniquevaluescount(data, 'mental_health_interview')

mental_health_interview  Count
                     No   1008
                  Maybe    207
                    Yes     44
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###phys_health_interview

Checking the unique values in the 'phys_health_interview' feature.

In [520]:
# Checking the unique values in the 'phys_health_interview' feature
uniquevaluescount(data, 'phys_health_interview')

phys_health_interview  Count
                Maybe    557
                   No    500
                  Yes    202
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###mental_vs_physical

Checking the unique values in the 'mental_vs_physical' feature.

In [521]:
# Checking the unique values in the 'mental_vs_physical' feature
uniquevaluescount(data, 'mental_vs_physical')

mental_vs_physical  Count
        Don't know    576
               Yes    343
                No    340
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###obs_consequence

Checking the unique values in the 'obs_consequence' feature.

In [522]:
# Checking the unique values in the 'obs_consequence' feature
uniquevaluescount(data, 'obs_consequence')

obs_consequence  Count
             No   1075
            Yes    184
Total : 1259


There seems no problem with the unique values: no typo or any abnormal format.

###comments

In [531]:
for i in data.columns[0:len(data.columns)]:
    print(uniquevaluescount(data, i))

 Age  Count
  29     85
  32     82
  26     75
  31     75
  27     71
  33     70
  28     68
  34     65
  30     63
  25     61
  35     55
  23     51
  24     46
  37     43
  38     39
  36     37
  39     33
  40     33
  43     28
  22     21
  41     21
  42     20
  21     16
  45     12
  46     12
  44     11
  19      9
  18      7
  20      6
  48      6
  50      6
  51      5
  49      4
  56      4
  54      3
  55      3
  57      3
  47      2
  60      2
  53      1
  58      1
  61      1
  62      1
  65      1
  72      1
Total : 1259
None
Gender  Count
  Male    985
Female    244
 Other     30
Total : 1259
None
               Country  Count
         United States    751
        United Kingdom    185
                Canada     72
               Germany     45
               Ireland     27
           Netherlands     27
             Australia     21
                France     13
                 India     10
           New Zealand      8
                 Italy    

The 'comments' feature has 161 unique values large missing values. We will check the percentage of the missing values.

In [523]:
# Checking the missing value percentage of 'comments' feature
missingvaluepercentage(data, 'comments')

Missing value : 86.97%


Since the feature has a large missing values, we will omit this feature.

In [524]:
# Dropping the 'comments' feature
data.drop('comments', axis = 1, inplace = True)

# Checking the features of the dataset
print(data.columns)

Index(['Age', 'Gender', 'Country', 'state', 'self_employed', 'family_history',
       'treatment', 'work_interfere', 'no_employees', 'remote_work',
       'tech_company', 'benefits', 'care_options', 'wellness_program',
       'seek_help', 'anonymity', 'leave', 'mental_health_consequence',
       'phys_health_consequence', 'coworkers', 'supervisor',
       'mental_health_interview', 'phys_health_interview',
       'mental_vs_physical', 'obs_consequence'],
      dtype='object')


The 'comments' feature has been successfully omitted while retaining the other features.

##**4. Data Exporting and Next Actions**

Now we have done cleaning and doing some feature engineering to the initial dataset. What is left is to export the result in a new csv. file.

In [525]:
data.to_csv('Mental Health Survey (Cleaned).csv')